assignments/this.js

```diff
@@ -1,26 +1,59 @@
 /* The for principles of "this";
 * in your own words. explain the four principle for the "this" keyword below.
 *
-* 1. 
-* 2. 
-* 3. 
-* 4. 
+* 1. Global Context - Java code executes in global context. When 'this' is called in global scope it will refer to global object.
+*  If we call 'this' from function`s scope, it will refer to global object as well. In this two cases our context is global scope.
+* 2. Implicit Binding - This is used for adding the same statements to different context.
+*  We can just add the function to needed context (object) and then this keyword will refer to the context. In this case our context is object`s scope
+* 3. Explicit Binding - (call, apply, bind) -
+* 4. New Binding -
 *
 * write out a code example of each explanation above
 */
 
 // Principle 1
-
+this
 // code example for Window Binding
-
+function globalContext() {
+    console.log(this);
+    return 'gav';
+}
+globalContext();
 // Principle 2
-
 // code example for Implicit Binding
-
+const sayHi = (obj) => {
+  obj.sayHello = function () {
+      console.log(`My context is ${this.species}`);
+      return `Hello, I am a ${this.species}`;
+  };
+};
+const dog = {species: 'dog'};
+const cat = {species: 'cat'};
+sayHi(dog);
+sayHi(cat);
+console.log(dog.sayHello());
+console.log(cat.sayHello());
 // Principle 3
-
 // code example for New Binding
-
+const startTheEngine = function(car1, car2) { return `The ${this.car_make} engine started ${car1} ${car2}`};
+const car = {car_make: 'Audi', car_model: 'Q7', car_year: 2009};
+const arrCar = ['ddd', 'dd'];
+console.log(startTheEngine.call(car, arrCar[0], arrCar[1]));
+console.log(startTheEngine.apply(car, arrCar));
+const audi = startTheEngine.bind(car, arrCar[0], arrCar[1]);
+console.log(audi());
```

Comment:
When a constructor function is actually used, or instantiated, then a 'new' instance of that object can be said to be made.  In these instances when a new object is constructed, or invoked, the 'this' keyword inside the constructor function above is bound to the new object. 

// code example for New Binding
const Chessplayer = function(name, whiteOpening, blackOpening) {
    this.name = name;
    this.whiteOpening = whiteOpening;
    this.blackOpening = blackOpening;
    this.speak = function() {
      console.log(`Hello, my name is ${this.name} and my favorite openings are the ${this.whiteOpening} and the ${this.blackOpening}`);
    }
};
// when a new object is constructed, or invoked, as below, the 'this' keyword inside the function above is bound to the new object
const challenger = new Chessplayer('Fabiano Caruana', 'Queens Gambit', 'French Defense');

challenger.speak();